### ISSUE TITLE
Improve DataLoader performance

### ISSUE THREAD

**Issue by tusharmath** [tailcallhq/tailcall#419]
It should be as fast as http-cache using moka. Currently batching makes the overall GraphQL server slower.

**Additional Context**
- Can consider using a lock-free hashmap via. a third-party crate.
- Here are some benchmarks to show the **`40%`** drop in performance when batching is enabled with the current implementation.
  
  **With Batching**
  
  ```graphql
  schema
    @server(port: 8000, enableGraphiql: true, enableQueryValidation: false)
    @upstream(baseURL: "http://jsonplaceholder.typicode.com", enableHttpCache: true, batch: {delay: 0}) {
    query: Query
  }
  
  type Query {
    posts: [Post] @http(path: "/posts")
  }
  
  type Post {
    id: Int!
    userId: Int!
    title: String!
    body: String!
  }
  ```
  
  
  
  ```
  ❯ wrk -t4 -c100 -d5s -s wrk.lua http://localhost:8000/graphql
  Running 5s test @ http://localhost:8000/graphql
    4 threads and 100 connections
    Thread Stats   Avg      Stdev     Max   +/- Stdev
      Latency    13.38ms   43.42ms 323.26ms   95.15%
      Req/Sec     6.37k     1.20k    7.66k    84.57%
    120885 requests in 5.01s, 620.46MB read
  Requests/sec:  24150.65
  Transfer/sec:    123.96MB
  ```
  
  
  **Without Batching**
  
  ```graphql
  schema
    @server(port: 8000, enableGraphiql: true, enableQueryValidation: false)
    @upstream(baseURL: "http://jsonplaceholder.typicode.com", enableHttpCache: true) {
    query: Query
  }
  
  type Query {
    posts: [Post] @http(path: "/posts")
  }
  
  type Post {
    id: Int!
    userId: Int!
    title: String!
    body: String!
  }
  ```
  
  
  ```
  ❯ wrk -t4 -c100 -d5s -s wrk.lua http://localhost:8000/graphql
  Running 5s test @ http://localhost:8000/graphql
    4 threads and 100 connections
    Thread Stats   Avg      Stdev     Max   +/- Stdev
      Latency     2.60ms    1.63ms  29.99ms   79.92%
      Req/Sec    10.05k     1.11k   10.94k    93.14%
    203947 requests in 5.10s, 1.02GB read
  Requests/sec:  39980.75
  Transfer/sec:    205.21MB
  ```

**Comment by artech-git** [comment]
> It should be as fast as http-cache using moka. Currently batching makes the overall GraphQL server slower.
> 
> **Additional Context**
> 
> * We only need to support batching capability
> * Can consider using a lock-free hashmap via. a third-party crate.

If Iam understanding this correct are you referring to `HttpDataLoader<C>` here? and more so specifically this ?

```rust
#[async_trait::async_trait]
impl<C: HttpClient + Send + Sync + 'static + Clone> Loader<DataLoaderRequest> for HttpDataLoader<C> {
...
```
I studied it for a while and felt that a concurrent lock-free HashMap here can speed up the performance by lot😃.

let me know about it more in detail.

**Comment by tusharmath** [comment]
@artech-git Yes, the currently using data loader degrades performance. Let me update the issue with the benchmarks.

**Comment by tusharmath** [comment]
@artech-git Updated the issue with the benchmarks, there is around a 40% degradation in performance using the current data-loader. Data Loader should technically degrade performance but not by this much.

**Comment by tusharmath** [comment]
Let me know if you would like to pick this up, I can assign it to you :)

**Comment by artech-git** [comment]
> Let me know if you would like to pick this up, I can assign it to you :)

Sure :), I will do my best.

**Comment by artech-git** [comment]
Can you please elaborate more about "We only need to support batching capability" with some relevant examples.

**Comment by artech-git** [comment]
@tusharmath there is a line present in the `data_loader.rs` at line 46 which sorts all the keys first 
```rust
keys.sort_by(|a, b| ... );
...
for key in &key...
```
then performs some operation further, need to know why we have to sort these keys first ? 
It will be better with some basic examples :).

**Comment by tusharmath** [comment]
Good catch! I think it's for unit tests. I guess we could feature flag this and enable it only if there is a unit test running. For general usecases we don't need to sort.

**Comment by artech-git** [comment]
@tusharmath thank you so much for that information, I have been struggling since yesterday for `&[DataLoaderRequest]` to avoid heavy allocation during runtime 😅 I was about to resort using unsafe for that line.

**Comment by tusharmath** [comment]
Created this — https://github.com/tailcallhq/tailcall/issues/596
Let's come to it later, as this might require more changes.

**Comment by artech-git** [comment]
> Created this — #596 Let's come to it later, as this might require more changes.

I think maybe we should first try to test out the optimizations first and then move to configuring the sorting part?

**Comment by meskill** [comment]
I wonder if `{delay: 0}` is set on the purpose in the issue description? Is this has be done to if fact just to disable batching but to test batching branch of the code?

**Comment by tusharmath** [comment]
@meskill Yes, even with delay 0, batching is enabled. The requests are just put at the end of the internal task queue (AFAIK). It works within the context of a single client request, but not across multiple graphql requests.

**Comment by tusharmath** [comment]
We can close this for now. Fixes were made in #739
